---
title: "DE_genes"
author: "Jana"
date: '2022-05-02'
output: BiocStyle::html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
BiocStyle::html_document 

file_path <- "/Users/janabraunger/Documents/Studium/Master/Boston/Masterarbeit/"
figdir = paste0(file_path, "Figures/DE_genes/")
if(!dir.exists(figdir)) dir.create(figdir)
knitr::opts_chunk$set(dev = c("png", "pdf"), fig.path = figdir)
```

```{r load_libraries, message = FALSE, warning = FALSE}
library(tidyverse)
library(magrittr )
library(ggplot2)
library(ggfortify)
library(MOFA2)
library(dplyr)
library(MOFAdata)
library(GGally)
library(ggpubr)
library(pheatmap)
library(RColorBrewer)
library(biomaRt)
library(UpSetR)
```


# Load RNAseq data

```{r load_data}
data <- read.delim(paste0(file_path, "Data/GSE113957_fpkm.txt"), 
                                     sep = "\t", header = TRUE) %>%
  # filter protein coding transcripts
  filter(grepl('protein-coding', Annotation.Divergence)) 
data[1:3, 1:10]
dim(data)
```

# Data preprocessing
## Gene annotations
```{r}
mart <- useEnsembl(biomart = "ensembl", 
                   dataset = "hsapiens_gene_ensembl", 
                   mirror = "uswest")
genes <- data$Transcript.ID
gene_annotations <- getBM(filters="refseq_mrna", 
                          attributes=c("refseq_mrna", "ensembl_gene_id", 
                                       "chromosome_name", "start_position", 
                                       "end_position", "external_gene_name", "description", 
                                       "ensembl_peptide_id"), 
                          values=genes, mart=mart) %>%
  data.frame()
colnames(gene_annotations) <- c("refseq_id", "ensembl_id", "chr", "start", "end", "symbol", "description", "peptide_id")
head(gene_annotations, n = 3)
```


## Rename from Refseq ID to Ensembl ID
```{r }
data_ens <- data  %>%
  rename(refseq_id = Transcript.ID) %>%
  # add ensembl gene id 
  left_join(., subset(gene_annotations, select = c(refseq_id, symbol)), by = "refseq_id") %>%
  distinct() %>%
  filter(symbol != "") 
```


## Convert to long format
```{r}
data_long <- subset(data_ens, select = -c(chr, start, end, strand, Length, Copies, Annotation.Divergence, refseq_id)) %>%
  #convert to long format
  gather(key = "name", value = "value", -c("symbol")) %>%
  #split name column into information categories
  separate(name, c("id", "age", "sex", "nationality"), sep = "_") %>%
  #remove patients with HGPS (premature aging disease)
  filter(!grepl('HGPS', sex)) %>% 
  #remove months from age
  mutate(age = gsub("y[rs]\\d+mos", "", age)) %>%
  #remove "yr" in age column to have the age as a number in all samples
  mutate(age = str_replace_all(age, c("yr" = "", "YR" = ""))) %>%
  mutate(age = as.integer(age)) %>%
  #rename sex to "female" and "male"
  mutate(sex = str_replace_all(sex, c("Female" = "female", "F" = "female", 
                                      "Male" = "male", "M" = "male"))) %>%
  rename(gene = symbol) 
head(data_long, n = 3)
```


## Mean over all samples with the same age 
```{r}
#round age down to 10 year intervals
data_long %<>% mutate(age = plyr::round_any(age, 10, f = floor))

# get mean expression per age
data_mean <- group_by(data_long, gene, age) %>%
  summarize(mean = mean(value)) %>%
  mutate(age = as.integer(age)) %>%
  #sort rows by age
  arrange(age) 
head(data_mean, n = 3)
age <- unique(data_mean$age)
```


## Filter highly expressed genes
```{r}
selected_genes <- group_by(data_mean, gene) %>%
  summarize(max = max(mean)) %>%
  filter(max > 0.1)

print(paste0(nrow(selected_genes), " genes with FPKM > 0.1 in at least one age group were selected out of ", length(unique(data_mean$gene)), " genes in total."))

data_filtered <- filter(data_mean, gene %in% selected_genes$gene) %>%
  # add pseudo-count
  mutate(mean = mean + 1) 
```


```{r hist_FPKM_filtered}
## Visualize the FPKM distribution 
ggplot(data_filtered, aes(x = mean)) +
  geom_histogram(bins = 100) +
  scale_y_continuous(trans='log2') +
  xlab('FPKM') +
  ylab('log2(counts)')
```

```{r}
# conversion to wide format
data_filtered %<>% spread(key = age, value = mean)
head(data_filtered, n = 3)
```


# Selection of differentially expressed genes
## Calculate log2 fold change
```{r}
age_steps <- seq(0, 80, 10)
data_fc <- lapply(age_steps, function(t0) {
  t1 <- t0 + 10
  # calculate log2 foldchange between two age intervals 
  fc <- log2(data_filtered[, toString(t1)] / data_filtered[, toString(t0)])
  colnames(fc) <- "log2_fc"
  df <- tibble(gene = data_filtered$gene, 
               FPKM_t0 = unlist(data_filtered[, toString(t0)]),
               FPKM_t1 = unlist(data_filtered[, toString(t1)]), 
               fc, 
               time = paste0('fc_', t0, '_', t1))
  return(df)}) %>%
  bind_rows() %>%
  mutate(abs_log2_fc = abs(log2_fc))

head(data_fc, n = 3)
```


## Select 1% of genes with highest abs fc per time step
```{r}
# select 1% with highest fc per interval
quantiles = group_by(data_fc, time) %>%
     summarize(threshold = quantile(abs(log2_fc), 0.99))
data_fc %<>% left_join(quantiles) %>%
  mutate(selected = ifelse(abs_log2_fc > threshold, 'true', 'false')) %>%
  mutate(selected = factor(selected, levels = c('true', 'false')))

fc_top <- filter(data_fc, selected == 'true') %>%
  subset(select = c(time, gene, FPKM_t0, FPKM_t1, log2_fc, abs_log2_fc))
head(fc_top, n = 3)
```


# Visualizations of DE genes
## Distribution of fold changes
```{r hist_fc}
ggplot(data_fc, aes(x = abs(log2_fc), fill = selected)) + 
  geom_histogram(bins = 100) + 
  facet_wrap(~time) +
  theme_bw() + 
  xlab('absolute log2 fold change') +
  theme(text = element_text(size = 15)) +
  scale_fill_manual(values = c("true" = "red",
                                "false"="darkgrey"))
```


## Intersection of DE genes
```{r upset, fig.height = 7, fig.width = 10}
#split fc_top into list of groups
top_list <- fc_top %>% group_by(time) %>% group_split()
names(top_list) <- unique(fc_top$time)
top_list <- lapply(top_list, function(list) list$gene)

upset(fromList(top_list), order.by = "freq", nsets = 9,
       mainbar.y.label = "Intersections", sets.x.label = "DE Genes",
       text.scale = c(2, 2, 1.5, 1.5, 2, 2), nintersects = 20)
```


## Gene expression of following age intervals
```{r fpkm, fig.height = 6}
ggplot(data_fc, aes(x = log2(FPKM_t0 + 1), y = log2(FPKM_t1 + 1), color = selected)) + 
  geom_point(size = 0.5) + 
  facet_wrap(~time) + 
  theme_bw() +
  scale_color_manual(values = c("true" = "red",
                                "false"="darkgrey")) + 
  geom_smooth(method='lm', color = 'black', size = 0.3) +
  theme(text = element_text(size = 15))
```


# Saving results data frame
```{r}
write.csv(fc_top, paste0(file_path, "Data/fc_top1.csv"), row.names = FALSE)
```

