---
title: "Test_overlapping_intervals"
author: "Jana"
date: '2022-04-21'
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
BiocStyle::html_document 

file_path <- "/Users/janabraunger/Documents/Studium/Master/Boston/Masterarbeit/MOFA/"
figdir = paste0(file_path, "Figures/overlapping_intervals/")
if(!dir.exists(figdir)) dir.create(figdir)
knitr::opts_chunk$set(dev = c("png", "pdf"), fig.path = figdir)
```

```{r load_libraries, message = FALSE, warning = FALSE}
library(tidyverse)
library(magrittr )
library(ggplot2)
library(ggfortify)
library(MOFA2)
library(dplyr)
library(MOFAdata)
library(GGally)
library(ggpubr)
library(pheatmap)
library(RColorBrewer)
library(biomaRt)
```


# Load data
## RNAseq data
```{r load_data}
data <- read.delim(paste0(file_path, "Data/GSE113957_fpkm.txt"), 
                                     sep = "\t", header = TRUE) %>%
  # filter protein coding transcripts
  filter(grepl('protein-coding', Annotation.Divergence)) 
data[1:5, 1:10]
dim(data)
```


## Gene annotations
```{r}
mart <- useEnsembl(biomart = "ensembl", 
                   dataset = "hsapiens_gene_ensembl", 
                   mirror = "uswest")
genes <- data$Transcript.ID
gene_annotations <- getBM(filters="refseq_mrna", 
                          attributes=c("refseq_mrna", "ensembl_gene_id", 
                                       "chromosome_name", "start_position", 
                                       "end_position", "external_gene_name", "description"), 
                          values=genes, mart=mart) %>%
  data.frame()
colnames(gene_annotations) <- c("refseq_id", "ensembl_id", "chr", "start", "end", "symbol", "description")
head(gene_annotations)
```

## Rename from Refseq ID to Ensembl ID
```{r }
data_ens <- data  %>%
  rename(refseq_id = Transcript.ID) %>%
  # add ensembl gene id 
  left_join(., subset(gene_annotations, select = c(refseq_id, ensembl_id)), by = "refseq_id")
```


## Convert to long format
```{r}
data_long <- subset(data_ens, select = -c(chr, start, end, strand, Length, Copies, Annotation.Divergence, refseq_id)) %>%
  gather(key = "name", value = "value", -c("ensembl_id")) %>%
  separate(name, c("id", "age", "sex", "nationality"), sep = "_") %>% 
  #remove "yr" in age column to have the age as a number in all samples
  mutate(age = str_replace_all(age, c("yr" = "", "YR" = ""))) %>%
  #remove samples with age in months instead of years
  filter(!grepl('mos', age)) %>%
  mutate(age = as.integer(age)) %>%
  #remove patients with HGPS (premature aging disease)
  filter(!grepl('HGPS', sex))
head(data_long)
```

## Mean over overlapping age intervals and log-transformation
```{r}
interval = 20
ages = seq(0, 96 - interval, 1)

data_mean <- lapply(ages, function(age_start){
  # calculate mean per age interval
  filter(data_long, between(age, age_start, age_start + interval)) %>%
    group_by(ensembl_id) %>%
    summarize(mean = mean(value)) %>%
    mutate(age = age_start)
}) %>% 
  # combine list of dfs into single df
  bind_rows() %>%
  mutate(age = as.integer(age)) %>%
  #sort rows by age
  arrange(age) %>%
  #log transformation
  mutate(mean = log(mean + 1))

head(data_mean)
```


## Select most variable genes
```{r}
var <- group_by(data_mean, ensembl_id) %>%
  # get mean per gene over all ages
  summarize(variance = var(mean)) %>%
  arrange(desc(variance)) %>%
  top_n(2000)
head(var)
```

```{r}
filtered_data <- filter(data_mean, ensembl_id %in% var$ensembl_id)
```


## PCA as a quality control
```{r pca}
data_wide <- spread(filtered_data, key = age, value = mean)
head(data_wide)

df_pca <- ungroup(data_wide)%>%
  subset(select = -c(ensembl_id))%>%
  t() %>%
  data.frame()%>%
  mutate(age = colnames(data_wide[-1])) %>%
  mutate(age = factor(age, levels = age, order = TRUE))

pca <- prcomp(df_pca[, 1:ncol(df_pca)-1])
autoplot(pca, data = df_pca, colour = "age")  + 
  theme(text = element_text(size = 20))
```

# MEFISTO
## Create MEFISTO object
```{r}
df_mefisto <- subset(filtered_data, select = c("age", "ensembl_id", "mean")) %>%
  mutate(view = "single_view") 
colnames(df_mefisto) <- c("sample", "feature", "value", "view")
head(df_mefisto)

# df with the corresponding covariate (age) for each sample 
time <- data.frame(sample = ages, covariate = "age", value =  ages)
```

```{r}
mefisto <- create_mofa(df_mefisto)
mefisto <- set_covariates(mefisto, covariates = time)
print(mefisto)
```

## Define options
```{r}
data_opts <- get_default_data_options(mefisto)

model_opts <- get_default_model_options(mefisto)
model_opts$num_factors <- 5

train_opts <- get_default_training_options(mefisto)

mefisto_opts <- get_default_mefisto_options(mefisto)

mefisto <- prepare_mofa(mefisto, model_options = model_opts,
                   mefisto_options = mefisto_opts,
                   training_options = train_opts,
                   data_options = data_opts)
```


## Build and train the MOFA object
```{r}
outfile = (paste0(file_path, "models/test_model_overlap.hdf5"))
trained_model <- run_mofa(mefisto, outfile)
```

# Downstream Analysis
```{r variance_explained}
plot_variance_explained(trained_model)
```

```{r}
get_scales(trained_model)
```


```{r factors_time}
plot_factors_vs_cov(trained_model) + 
  geom_line()
```


```{r heatmap_genes_highest_weight, fig.width=10}
plot_data_heatmap(trained_model,
  view = "single_view",         # view of interest
  factor = 1,             # factor of interest
  features = 15,          # number of features to plot (they are selected by weight)
  
  # extra arguments that are passed to the `pheatmap` function
  cluster_rows = TRUE, cluster_cols = TRUE,
  show_rownames = TRUE, show_colnames = TRUE)
```


# GSEA with reactome gene sets
## Load reactomeGS
```{r}
utils::data("reactomeGS")
head(rownames(reactomeGS))
head(colnames(reactomeGS))
```

 
## Run GSEA on positive weights
```{r}
enrichment.parametric <- run_enrichment(trained_model,
  view = "all", factors = 1:5,
  feature.sets = reactomeGS,
  sign = "positive",
  statistical.test = "parametric"
)
```


**Factor 1**
```{r gsea_reactome_f1_pos}
plot_enrichment(enrichment.parametric, 
  factor = 1, 
  max.pathways = 5) +
  theme_classic(base_size = 13)
```

## Run GSEA on negative weights
```{r}
enrichment.parametric <- run_enrichment(trained_model,
  view = "all", factors = 1:5,
  feature.sets = reactomeGS,
  sign = "negative",
  statistical.test = "parametric"
)
```

**Factor 1**
```{r gsea_reactome_f1_neg}
plot_enrichment(enrichment.parametric, 
  factor = 1, 
  max.pathways = 10) +
  theme_classic(base_size = 13)
```


# GSEA with MSigDB gene sets
## Run GSEA on weights with positive sign
```{r}
utils::data("MSigDB_v6.0_C5_human") 
enrichment.parametric <- run_enrichment(trained_model,
  view = "all", factors = 1:5,
  feature.sets = MSigDB_v6.0_C5_human,
  sign = "positive",
  statistical.test = "parametric"
)
```


**Factor 1**
```{r gsea_msig_f1_pos}
plot_enrichment(enrichment.parametric, 
  factor = 1, 
  max.pathways = 5) +
  theme_classic(base_size = 13)
```

## Run GSEA on weights with positive sign
```{r}
utils::data("MSigDB_v6.0_C5_human") 
enrichment.parametric <- run_enrichment(trained_model,
  view = "all", factors = 1:5,
  feature.sets = MSigDB_v6.0_C5_human,
  sign = "negative",
  statistical.test = "parametric"
)
```


**Factor 1**
```{r gsea_msig_f1_neg}
plot_enrichment(enrichment.parametric, 
  factor = 1, 
  max.pathways = 5) +
  theme_classic(base_size = 13)
```


# Genes with highest weights

## Get weight matrix
```{r}
weights <- get_weights(trained_model, views = "all", factors = "all", scale = F, as.data.frame = TRUE) %>%
  group_by(factor, view) %>% 
  mutate(value = value/max(abs(value))) %>% 
  ungroup() 
colnames(weights) <- c("ensembl_id", "factor", "value", "view")
weights <- weights[order(abs(weights$value), decreasing = TRUE), ] %>%
  left_join(subset(gene_annotations, select = c(ensembl_id, symbol, description)))
head(weights)
```

## Genes with highest weights in factor 1
**Select the six genes with highest absolute weight in factor 1**
```{r}
selected_genes <- filter(weights, factor == "Factor1") %>%
  top_n(6, abs(value))
selected_genes
```

**Plot expression of selected genes during aging**
```{r f1_expression_over_time}
df_plot = data_mean %>%
  left_join(subset(gene_annotations, select = c(ensembl_id, symbol))) %>%
  filter(ensembl_id %in% selected_genes$ensembl_id)

ggplot(df_plot, aes(x = age, y = mean, col = age)) + 
    geom_point() + 
    geom_line() +
    facet_wrap(~ symbol, scales = "free") +
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
    labs(y = "normalized gene expression (log FPKM)")
```




**Distribution of weights in factor 1**

```{r}
weights_f1 <- filter(weights, factor == "Factor1")
ggplot(weights_f1, aes(x = value)) + 
  geom_histogram(bins = 100)
```

**Save 10% of genes with highest positive and negative weight in factor 1**
```{r}
# Select genes with highest positive weight
top_genes_pos <- weights_f1 %>%
  top_frac(0.1, value) %>%
  left_join(subset(gene_annotations, select = c(ensembl_id, chr, start, end)), by = "ensembl_id") %>%
  mutate(locus = paste0("chr_", chr, "_loc_", 
                        plyr::round_any(start, 250000, floor))) %>%
  subset(select = c(ensembl_id, value, symbol, locus))

# Select genes with highest negative weight
top_genes_neg <- weights_f1 %>%
  top_frac(0.1, -value) %>%
  left_join(subset(gene_annotations, select = c(ensembl_id, chr, start, end)), by = "ensembl_id") %>%
  mutate(locus = paste0("chr_", chr, "_loc_", 
                        plyr::round_any(start, 250000, floor))) %>%
  subset(select = c(ensembl_id, value, symbol, locus))

head(top_genes_pos)

# Save as csv
#write.csv(top_genes_pos, paste0(file_path, "Data/top_genes_pos.csv"), row.names = FALSE)
#write.csv(top_genes_neg, paste0(file_path, "Data/top_genes_neg.csv"), row.names = FALSE)
```

## Comparison with active loci in IMR90 and GM23248
```{r}
IMR90 <- read.csv(paste0(file_path, "Data/IMR90_cosine_all_features.txt")) %>%
  rename(locus = X0) %>%
  mutate(young_activity = "active")
old_fibroblasts <- read.csv(paste0(file_path, "Data/old_fibroblasts_cosine_all_features.txt")) %>%
  rename(locus = X0) %>%
  mutate(old_activity = "active")

# add for each gene whether corresponding loci is active or inactive based on the HiC clustering in IMR90 and GM23248
activity <- top_genes_pos %>%
  left_join(IMR90) %>% 
  replace_na(list(young_activity = "inactive")) %>%
  left_join(old_fibroblasts) %>% 
  replace_na(list(old_activity = "inactive")) %>%
  mutate(development = factor(paste0(young_activity, "_", old_activity), 
        levels = c("active_active", "active_inactive", "inactive_active", "inactive_inactive")))
head(activity)
```

```{r development_activity}
ggplot(activity, aes(x = development, fill = development)) +
  geom_bar()
```



