---
title: "Test_MEFISTO_sex"
author: "Jana"
date: '2022-04-21'
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
BiocStyle::html_document 

file_path <- "/Users/janabraunger/Documents/Studium/Master/Boston/Masterarbeit/MOFA/"
figdir = paste0(file_path, "Figures_sex/")
if(!dir.exists(figdir)) dir.create(figdir)
knitr::opts_chunk$set(dev = c("png", "pdf"), fig.path = figdir)
```

```{r load_libraries, message = FALSE, warning = FALSE}
library(tidyverse)
library(magrittr )
library(ggplot2)
library(ggfortify)
library(MOFA2)
library(dplyr)
library(MOFAdata)
library(GGally)
library(ggpubr)
library(pheatmap)
library(RColorBrewer)
library(biomaRt)
library(ggvenn)
```


# Data preprocessing
## RNAseq data
```{r load_data}
data <- read.delim(paste0(file_path, "Data/GSE113957_fpkm.txt"), 
                                     sep = "\t", header = TRUE) %>%
  # filter protein coding transcripts
  filter(grepl('protein-coding', Annotation.Divergence)) 
data[1:5, 1:10]
dim(data)
```


## Gene annotations
```{r}
mart <- useEnsembl(biomart = "ensembl", 
                   dataset = "hsapiens_gene_ensembl", 
                   mirror = "uswest")
genes <- data$Transcript.ID
gene_annotations <- getBM(filters="refseq_mrna", 
                          attributes=c("refseq_mrna", "ensembl_gene_id", 
                                       "chromosome_name", "start_position", 
                                       "end_position", "external_gene_name", "description"), 
                          values=genes, mart=mart) %>%
  data.frame()
colnames(gene_annotations) <- c("refseq_id", "ensembl_id", "chr", "start", "end", "symbol", "description")
head(gene_annotations)
```

## Rename from Refseq ID to Ensembl ID
```{r }
data_ens <- data  %>%
  rename(refseq_id = Transcript.ID) %>%
  # add ensembl gene id 
  left_join(., subset(gene_annotations, select = c(refseq_id, ensembl_id)), by = "refseq_id")
```


## Convert to long format
```{r}
data_long <- subset(data_ens, select = -c(chr, start, end, strand, Length, Copies, Annotation.Divergence, refseq_id)) %>%
  #convert to long format
  gather(key = "name", value = "value", -c("ensembl_id")) %>%
  #split name column into information categories
  separate(name, c("id", "age", "sex", "nationality"), sep = "_") %>%
  #remove patients with HGPS (premature aging disease)
  filter(!grepl('HGPS', sex)) %>% 
  #remove months from age
  mutate(age = gsub("y[rs]\\d+mos", "", age)) %>%
  #remove "yr" in age column to have the age as a number in all samples
  mutate(age = str_replace_all(age, c("yr" = "", "YR" = ""))) %>%
  mutate(age = as.integer(age)) %>%
  #rename sex to "female" and "male"
  mutate(sex = str_replace_all(sex, c("Female" = "female", "F" = "female", 
                                      "Male" = "male", "M" = "male")))
head(data_long)
```


## Visualize number of individuals per age
```{r hist_ages, fig.height=4}
samples <- subset(data_long, select = c(id, age, sex, nationality)) %>%
  distinct()
print(paste0('The dataset includes RNAseq samples from ', nrow(samples), ' patients.'))

ggplot(samples, aes(x = age, group = sex, fill = sex)) +
  geom_bar() +
  theme_bw() + 
  ylab('number of individuals') +
  theme(text = element_text(size = 17))
```


## Bin age into 10 years intervals
```{r hist_age_intervals, fig.height=4}
#round age down to 10 year intervals
data_long %<>% mutate(age = plyr::round_any(age, 10, f = floor))

#plot age distribution over the intervals
samples_binned <- subset(data_long, select = c(id, age, sex, nationality)) %>%
  distinct()
ggplot(samples_binned, aes(x = age, group = sex, fill = sex)) +
  geom_bar() +
  theme_bw() + 
  xlab('age intervals') +
  ylab('number of individuals') +
  theme(text = element_text(size = 17))
```


## Mean over all samples with the same age and log-transform
```{r}
# get mean expression per age
data_mean <- group_by(data_long, ensembl_id, age, sex) %>%
  summarize(mean = mean(value)) %>%
  mutate(age = as.integer(age)) %>%
  #sort rows by age
  arrange(age) %>%
  #log transformation
  mutate(mean = log(mean + 1))
head(data_mean)
age <- unique(data_mean$age)
```

## Select most variable genes
```{r}
var <- group_by(data_mean, ensembl_id) %>%
  # get var per gene over all ages
  summarize(variance = var(mean)) %>%
  arrange(desc(variance)) %>%
  top_n(2000)
head(var)
```

```{r}
filtered_data <- filter(data_mean, ensembl_id %in% var$ensembl_id)
```


# MEFISTO
## Create MEFISTO object
```{r}
df_mefisto <- subset(filtered_data, select = c("age", "ensembl_id", "mean", "sex")) 
colnames(df_mefisto) <- c("sample", "feature", "value", "view")
head(df_mefisto)

# df with the corresponding covariate (age) for each sample 
time <- data.frame(sample = age, covariate = "age", value =  age)
```

```{r}
mefisto <- create_mofa(df_mefisto)
mefisto <- set_covariates(mefisto, covariates = time)
print(mefisto)
```

## Define options
```{r}
data_opts <- get_default_data_options(mefisto)

model_opts <- get_default_model_options(mefisto)
model_opts$num_factors <- 3

train_opts <- get_default_training_options(mefisto)

mefisto_opts <- get_default_mefisto_options(mefisto)

mefisto <- prepare_mofa(mefisto, model_options = model_opts,
                   mefisto_options = mefisto_opts,
                   training_options = train_opts,
                   data_options = data_opts)
```


## Build and train the MOFA object
```{r}
outfile = (paste0(file_path, "models/test_model_sex.hdf5"))
trained_model <- run_mofa(mefisto, outfile)
```

# Downstream Analysis
```{r variance_explained}
plot_variance_explained(trained_model)
```

```{r}
get_scales(trained_model)
```

```{r factors_time}
plot_factors_vs_cov(trained_model) + 
  geom_line()+
  theme_bw()+ 
  theme(text = element_text(size = 15))
```
```{r heatmap_genes_highest_weight, fig.width=10}
plot_data_heatmap(trained_model,
  view = "male",         # view of interest
  factor = 1,             # factor of interest
  features = 15,          # number of features to plot (they are selected by weight)
  
  # extra arguments that are passed to the `pheatmap` function
  cluster_rows = TRUE, cluster_cols = TRUE,
  show_rownames = TRUE, show_colnames = TRUE, fontsize = 15) 
```


# Genes with highest weights

## Get weight matrix
```{r}
weights <- get_weights(trained_model, views = "all", factors = "all", scale = F, as.data.frame = TRUE) %>%
  group_by(factor, view) %>% 
  # normalize weights per factor and view
  mutate(value = value/max(abs(value))) %>% 
  ungroup() %>%
  # remove sex suffix from features
  mutate(feature = str_replace_all(feature, c("_female" = "", "_male" = ""))) 
colnames(weights) <- c("ensembl_id", "factor", "value", "view")
weights <- weights[order(abs(weights$value), decreasing = TRUE), ] %>%
  left_join(subset(gene_annotations, select = c(ensembl_id, symbol, description)))
head(weights)
```


## Genes with highest weights in factor 1
### Select the six genes with highest absolute weight in factor 1
```{r}
selected_genes_pos <- filter(weights, factor == "Factor1") %>%
  top_n(7, value)
selected_genes_neg <- filter(weights, factor == "Factor1") %>%
  top_n(7, -value)
selected_genes <- bind_rows(selected_genes_pos, selected_genes_neg)

selected_genes
```

### Plot expression of selected genes during aging
```{r f1_expression_over_time, fig.width=10, fig.height = 5}
df_plot = data_mean %>%
  left_join(subset(gene_annotations, select = c(ensembl_id, symbol))) %>%
  filter(ensembl_id %in% selected_genes$ensembl_id) %>%
  mutate(symbol = factor(symbol, levels = selected_genes$symbol))

ggplot(df_plot, aes(x = age, y = mean, col = sex)) + 
    geom_point() + 
    geom_line() +
    facet_wrap(~ symbol, scales = "free", ncol = 7) +
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
    labs(y = "normalized gene expression (log FPKM)")  +
    theme_bw() + 
    theme(text = element_text(size = 15))
```


### Overlap between males and females
```{r venn_f_m}
top_male <- filter(weights, factor == "Factor1") %>%
  filter(view == "male") %>% 
  top_n(200, abs(value))

top_female <- filter(weights, factor == "Factor1") %>%
  filter(view == "female") %>% 
  top_n(200, abs(value))

ggvenn(list(male = top_male$ensembl_id, female = top_female$ensembl_id))
```


```{r weights_f_m}
df <- distinct(filter(weights, factor == "Factor1")) %>%
  spread(key = view, value = value) 
  
ggplot(df, aes(x = male, y = female)) + 
  geom_point()
```






