---
title: "06_clustering_expression_patterns"
author: "Jana Braunger"
date: '2022-07-07'
output: BiocStyle::html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
BiocStyle::html_document 

dir <- "/Users/janabraunger/Documents/Studium/Master/Boston/Masterarbeit/"
figdir = paste0(dir, "Figures/expression_patterns/")
if(!dir.exists(figdir)) dir.create(figdir)
knitr::opts_chunk$set(dev = c("png", "pdf"), fig.path = figdir)
```

```{r load_libraries, message = FALSE, warning = FALSE}
library(tidyverse)
library(magrittr)
library(pheatmap)
```

# Load data
## Normalized counts
```{r}
counts <- read.delim(paste0(dir, "Data/count_data.csv"), sep = ",")
mean_counts <- read.delim(paste0(dir, "Data/mean_counts.csv"), sep = ",") %>%
  group_by(gene, age_group) %>%
  summarize(counts = mean(counts))
head(mean_counts)
```

## TF-target interactions
```{r}
TF_targets <- read.delim(paste0(dir, "Data/tf-target-information.txt")) %>%
  subset(select = c("TF", "target")) %>%
  distinct() %>%
  drop_na()
head(TF_targets)
```

## Differentially expressed genes
```{r}
p_0.05 <- read.delim(paste0(dir, "Data/DE_var_p_n_200.csv"), sep = ",")
head(p_0.05)
```

# Hierarchical clustering
## Select DE targets of DE TFs
```{r}
# data frame to map to the following transition
transitions <- data.frame(TF_transition = c("fc_1-15_16-26", "fc_16-26_27-60", "fc_27-60_61-85"), 
                          target_transition = c("fc_16-26_27-60", "fc_27-60_61-85", "fc_61-85_86-96"))

# select DE TFs and their DE targets in the following transition
DE_TFs_targets <- filter(p_0.05, gene %in% TF_targets$TF) %>%
  filter(transition != "fc_61-85_86-96") %>%
  dplyr::rename(TF = gene, TF_transition = transition, TF_fc = abs_log2_fc) %>%
  left_join(transitions) %>%
  left_join(TF_targets) %>%
  left_join(dplyr::rename(p_0.05, target = gene, target_transition = transition)) %>%
  drop_na(abs_log2_fc)

head(DE_TFs_targets)
group_by(DE_TFs_targets, TF) %>% summarize(n_DE_targets = n()) %>% 
  arrange(desc(n_DE_targets))
```


## Expression of targets
```{r}
expr_DE_targets <- left_join(DE_TFs_targets, dplyr::rename(mean_counts, target = gene)) %>%
  filter(TF == "FOS") %>%
  subset(select = c(target, age_group, counts)) %>%
  #drop_na() %>%
  spread(key = age_group, value = counts) 
head(expr_DE_targets)
```


## Calculate distance and perform clustering 
```{r dendrogram}
# Create matrix
hclust_matrix <- select(expr_DE_targets, -target) %>%
  as.matrix()
rownames(hclust_matrix) <- expr_DE_targets$target

# Z-transform values for each gene
hclust_matrix %<>% t() %>% scale() %>% t()

# Calculate euclidean distances between genes
gene_dist <- dist(hclust_matrix)

# Clustering
gene_hclust <- hclust(gene_dist, method = "ward.D2")

# Plot dendrogram
plot(gene_hclust, labels = FALSE)
```


# Gene expression trends per cluster
```{r expr_patterns_FOS_clustering, fig.height=4}
gene_cluster <- cutree(gene_hclust, k = 4) %>% 
  enframe() %>% 
  dplyr::rename(gene = name, cluster = value) %>%
  left_join(group_by(mean_counts, gene) %>% mutate(counts = scale(counts)))

ggplot(gene_cluster, aes(x = age_group, y = counts, group = gene)) +
  geom_line(alpha = 0.3) +
  facet_wrap(~cluster, nrow = 1)+
  geom_line(stat = "summary", fun = "median", colour = "brown", size = 1.2, aes(group = 1)) +
  theme_bw()
```

## Save clustering results
```{r}
results <- subset(gene_cluster, select = c(gene, cluster)) %>% distinct()
write.csv(results, paste0(dir, "Data/cluster_FOS_targets.csv"), row.names = FALSE)
```


# Difference map: young - old fibroblasts for FOS DE targets
```{r}
# Read in difference map for FOS DE targets
diff <- read.delim(paste0(dir, "Data/difference_maps/diff_FOS_DE_targets.csv"), sep = ",") %>%
  gather(key = 'chr2', value = 'difference', -chr1)

# Number of differences
diff_counts <- group_by(diff, chr1, difference) %>% tally() %>% filter(difference != 0) %>%
  left_join(data.frame("difference" = c(1, -1), "group" = c('young-specific', 'old-specific'))) %>%
  dplyr::rename('locus' = 'chr1') %>% 
  subset(select = c(locus, n, group)) %>%
  filter(n >= 7)
head(diff_counts)
```


```{r expr_patterns_FOS_age}
# Map loci to genes
loci <- unique(diff$chr1)
gene_loci <- read.delim(paste0(dir, "Data/all_gene_loci.csv"), sep = ",") %>%
  filter(gene %in% expr_DE_targets$target) %>%
  left_join(diff_counts) %>% 
  mutate(group = replace_na(group, "no_differences"))

# Add expression values
expr_targets <- left_join(gene_loci, mean_counts) %>%
  subset(select = c(gene, age_group, counts, group)) %>%
  group_by(gene) %>%
  mutate(counts = scale(counts))

ggplot(expr_targets, aes(x = age_group, y = counts, group = gene)) +
  geom_line(alpha = 0.3) +
  facet_wrap(~group, nrow = 1)+
  geom_line(stat = "summary", fun = "median", colour = "brown", size = 1.2, aes(group = 1)) +
  theme_bw()
```


# Difference map for loci from last PCST network
## Expression patterns for difference groups
```{r expr_pattern_last_PCST_age}
# Read in difference map for last PCST loci and Louvain clusters
diff <- read.delim(paste0(dir, "Data/difference_maps/diff_last_PCST_loci.csv"), sep = ",") %>%
  gather(key = 'chr2', value = 'difference', -chr1)
clusters <- read.delim(paste0(dir, 'Data/difference_maps/louvain_clusters_last_PCST.csv'), sep = ",") 
#clusters %<>% mutate(cluster = sample(clusters$cluster, nrow(clusters)))

# Number of differences
diff_counts <- group_by(diff, chr1, difference) %>% tally() %>% filter(difference != 0) %>%
  left_join(data.frame("difference" = c(1, -1), "group" = c('young-specific', 'old-specific'))) %>%
  dplyr::rename('locus' = 'chr1') %>% 
  subset(select = c(locus, n, group)) %>%
  filter(n >= 40)

# Add number of differences to the annotation df
gene_loci <- clusters %>%
  left_join(diff_counts) %>% 
  mutate(group = replace_na(group, "no_differences"))

# Add expression values
expr_targets <- left_join(gene_loci, mean_counts) %>%
  subset(select = c(gene, locus, group, cluster, age_group, counts)) %>%
  group_by(gene) %>%
  mutate(counts = scale(counts))

ggplot(expr_targets, aes(x = age_group, y = counts, group = gene)) +
  geom_line(alpha = 0.3) +
  facet_wrap(~group, nrow = 1)+
  geom_line(stat = "summary", fun = "median", colour = "brown", size = 1.2, aes(group = 1)) +
  theme_bw()
```


## Expression patterns per Louvain cluster
```{r PCST_lou_cluster }
ggplot(expr_targets, aes(x = age_group, y = counts, group = gene)) +
  geom_line(alpha = 0.3) +
  facet_wrap(~cluster, nrow = 2)+
  geom_line(stat = "summary", fun = "median", colour = "brown", size = 1.2, aes(group = 1)) +
  theme_bw()+
  theme(text = element_text(size = 15))
```


```{r PCST_lou_cluster_top}
# Number of differences
diff_counts <- group_by(diff, chr1, difference) %>% tally() %>% filter(difference != 0) %>%
  left_join(data.frame("difference" = c(1, -1), "group" = c('young-specific', 'old-specific'))) %>%
  dplyr::rename('locus' = 'chr1') %>% 
  subset(select = c(locus, n, group))

# Get top 10 loci per cluster (with most differences)
selection <- clusters %>%
  left_join(diff_counts) %>% 
  group_by(cluster) %>% 
  slice_max(n, n = 10, with_ties = FALSE)

# Expression patterns only for the top 10 loci per cluster
expr_targets_filt <- left_join(selection, mean_counts) %>%
  subset(select = c(gene, locus, group, cluster, age_group, counts)) %>%
  group_by(gene) %>%
  mutate(counts = scale(counts))

ggplot(expr_targets_filt, aes(x = age_group, y = counts, group = gene)) +
  geom_line(alpha = 0.3) +
  facet_wrap(~cluster)+
  geom_line(stat = "summary", fun = "median", colour = "brown", size = 1.2, aes(group = 1)) +
  theme_bw()+
  theme(text = element_text(size = 15))
```


## Distribution of differences among clusters 
```{r PCST_n_differences_clustering, fig.width = 10}
clusters_diff <- group_by(diff, chr1, difference) %>% tally() %>% filter(difference != 0) %>%
  left_join(data.frame("difference" = c(1, -1), "group" = c('young-specific', 'old-specific'))) %>%
  dplyr::rename('locus' = 'chr1') %>%
  left_join(clusters) %>%
  mutate(group = factor(group, levels = c('young-specific', 'old-specific'))) %>%
  group_by(group, cluster) %>%
  mutate(mean = mean(n))

ggplot(clusters_diff, aes(x = n)) + 
  geom_histogram() +
  facet_grid(rows = vars(group), cols = vars(cluster), scale = "free_x") +
  geom_vline(aes(xintercept = mean, group = cluster), colour = 'red') +
  theme_bw() +
  xlab('number of differences') +
  ylab('number of loci')+
  theme(text = element_text(size = 15))

```


## Distribution of differences per locus
```{r PCST_scatter_diff_lou_cluster}
clusters_diff_wide <- subset(clusters_diff, select = c(locus, gene, cluster, group, n)) %>%
  spread(key = group, value = n, fill = 0) %>%
  mutate(cluster = factor(cluster, levels = seq(0, 5, 1)))
colnames(clusters_diff_wide) <- c('locus', 'gene', 'cluster', 'young_specific', 'old_specific')

ggplot(clusters_diff_wide, aes(x = young_specific, y = old_specific, color = cluster)) +
  geom_point() +
  xlab('young-specific interactions per locus') +
  ylab('old-specific interactions per locus') +
  facet_wrap(~cluster) +
  theme_bw()+
  theme(text = element_text(size = 17)) +
  geom_vline(xintercept = 10) +
  geom_hline(yintercept = 10)
```


## Loci with many young and old specific interactions
```{r}
switching_loci <- filter(clusters_diff_wide, young_specific > 10, old_specific > 10)
```


## Clustering according to expression values
```{r PCST_dend}
# Create matrix
hclust_matrix <- select(ungroup(expr_targets), -c(group, locus, cluster)) %>%
  spread(key = age_group, value = counts) %>%
  column_to_rownames('gene') %>%
  as.matrix()

# Calculate euclidean distances between genes
gene_dist <- dist(hclust_matrix)

# Clustering
gene_hclust <- hclust(gene_dist, method = "ward.D2")

# Plot dendrogram
plot(gene_hclust, labels = FALSE)
```

```{r PCST_clustering}
gene_cluster <- cutree(gene_hclust, k = 6) %>% 
  enframe() %>% 
  dplyr::rename(gene = name, cluster = value) %>%
  left_join(group_by(mean_counts, gene) %>% mutate(counts = scale(counts)))

ggplot(gene_cluster, aes(x = age_group, y = counts, group = gene)) +
  geom_line(alpha = 0.3) +
  facet_wrap(~cluster, nrow = 2)+
  geom_line(stat = "summary", fun = "median", colour = "brown", size = 1.2, aes(group = 1)) +
  theme_bw()+
  theme(text = element_text(size = 15))
```


## TF enrichment per Louvain cluster
```{r}
targets_per_cluster <- left_join(clusters, dplyr::rename(TF_targets, 'gene' = 'target')) %>%
  group_by(TF, cluster) %>%
  summarize(number_of_targets = n()) %>%
  drop_na()
head(targets_per_cluster)
```

```{r TF_cluster_enrichment}
# for normalization: number of genes per cluster and percentages that each TF targets
genes_per_cluster <- group_by(clusters, cluster) %>% summarize(n = n())
# percentage of targets per TF
percentages <- targets_per_cluster %>%
  group_by(TF) %>%
  summarize(number_of_targets = sum(number_of_targets)) %>%
  mutate(percentage_targets = number_of_targets / length(clusters$gene))

wide <- spread(targets_per_cluster, key = cluster, value = number_of_targets, fill = 0) %>%
  column_to_rownames('TF')
colnames(wide) <- c('cluster_0', 'cluster_1', 'cluster_2', 'cluster_3', 'cluster_4', 'cluster_5')

# enrichment score: percentage of genes targeted in the cluster / percentages of genes targeted overall (in the PCST genes)
wide <- t(t(wide)/genes_per_cluster$n) 
#wide <- wide / percentages$percentage_targets
  
#pheatmap(wide, show_rownames = F, clustering_method = 'ward.D2', 
#         color = colorRampPalette(brewer.pal(n = 7, name = "YlOrRd"))(100))
pheatmap(wide, show_rownames = F, clustering_method = 'ward.D2', scale = "row")
```

--> There are groups of TFs that only target genes in one cluster specifically. 
Do these TF groups overlap with the ones we identified in our PCST analysis?  


```{r TF_cluster_enrichment_anno}
# load info about which TFs occurred in which of the three network
pcst_TFs = read.delim(paste0(dir, 'Data/pcst/incl_TFs_design2.csv'), sep = ",") %>%
  mutate(included = "True") %>%
  mutate(net = factor(net, levels = c('young_net', 'middle_net', 'old_net'))) %>% #keeps ordering
  spread(key = net, value = included, fill = "False") %>%
  right_join(data.frame('TF' = rownames(wide))) %>%
  replace(is.na(.), 'False') %>%
  column_to_rownames('TF')
head(pcst_TFs)

# add annotation bars according to PCST
annoCol<-list(young_net=c(True="blue", False="grey"), 
              middle_net = c(True = 'red', False = 'grey'), 
              old_net = c(True = 'green', False = 'grey'))

pheatmap(wide, show_rownames = F, clustering_method = 'average',
         annotation_row = pcst_TFs, annotation_colors = annoCol, scale = "row")
```




