---
title: "MEFISTO_number-factors"
author: "Jana"
date: '2022-05-09'
output: BiocStyle::html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
BiocStyle::html_document 

# Set path
file_path <- "/Users/janabraunger/Documents/Studium/Master/Boston/Masterarbeit/MOFA/"

# Set options for the figures
figdir = paste0(file_path, "Figures/n_factors")
if(!dir.exists(figdir)) dir.create(figdir)
knitr::opts_chunk$set(dev = c("png", "pdf"), fig.path = figdir)

# Not use exponential representation of numbers
options(scipen = 999) 
```


```{r load_libraries, message = FALSE, warning = FALSE}
library(tidyverse)
library(magrittr )
library(ggplot2)
library(ggfortify)
library(MOFA2)
library(dplyr)
library(MOFAdata)
library(GGally)
library(ggpubr)
library(pheatmap)
library(RColorBrewer)
library(biomaRt)
library(corrr)
```


# Data preprocessing
## RNAseq data
```{r load_data}
data <- read.delim(paste0(file_path, "Data/GSE113957_fpkm.txt"), 
                                     sep = "\t", header = TRUE) %>%
  # filter protein coding transcripts
  filter(grepl('protein-coding', Annotation.Divergence)) 
data[1:5, 1:10]
dim(data)
```


## Gene annotations
```{r}
mart <- useEnsembl(biomart = "ensembl", 
                   dataset = "hsapiens_gene_ensembl", 
                   mirror = "uswest")
genes <- data$Transcript.ID
gene_annotations <- getBM(filters="refseq_mrna", 
                          attributes=c("refseq_mrna", "ensembl_gene_id", 
                                       "chromosome_name", "start_position", 
                                       "end_position", "external_gene_name", "description"), 
                          values=genes, mart=mart) %>%
  data.frame() %>%
  group_by(refseq_mrna) %>%
  summarise(ensembl_id = first(ensembl_gene_id), 
            chr = first(chromosome_name),
            start = first(start_position), 
            end = first(end_position),
            symbol = first(external_gene_name), 
            description = first(description))

head(gene_annotations)
```

## Rename from Refseq ID to Ensembl ID
```{r }
data_ens <- data  %>%
  rename(refseq_mrna = Transcript.ID) %>%
  # add ensembl gene id 
  left_join(., subset(gene_annotations, select = c(refseq_mrna, ensembl_id)), by = "refseq_mrna") 
```


## Convert to long format
```{r}
data_long <- subset(data_ens, select = -c(chr, start, end, strand, Length, Copies, Annotation.Divergence, refseq_mrna)) %>%
  #convert to long format
  gather(key = "name", value = "value", -c("ensembl_id")) %>%
  #split name column into information categories
  separate(name, c("id", "age", "sex", "nationality"), sep = "_") %>%
  #remove patients with HGPS (premature aging disease)
  filter(!grepl('HGPS', sex)) %>% 
  #remove months from age
  mutate(age = gsub("y[rs]\\d+mos", "", age)) %>%
  #remove "yr" in age column to have the age as a number in all samples
  mutate(age = str_replace_all(age, c("yr" = "", "YR" = ""))) %>%
  mutate(age = as.integer(age)) %>%
  #rename sex to "female" and "male"
  mutate(sex = str_replace_all(sex, c("Female" = "female", "F" = "female", 
                                      "Male" = "male", "M" = "male")))  %>%
  #mean over values from that come from different Refseq-IDs, but map to the same ensembl id
  group_by(ensembl_id, id, age, sex, nationality) %>%
  summarize(value = mean(value)) %>%
  ungroup() 

#round age down to 10 year intervals
data_long %<>% mutate(age = plyr::round_any(age, 10, f = floor))
head(data_long)
```


## Mean over all samples with the same age and log-transform the values
```{r}
# get mean expression per age
data_mean <- group_by(data_long, ensembl_id, age) %>%
  summarize(mean = mean(value)) %>%
  mutate(age = as.integer(age)) %>%
  #sort rows by age
  arrange(age) %>%
  #log transformation
  mutate(mean = log(mean + 1))
head(data_mean)
age <- unique(data_mean$age)
```


## Filter highly expressed genes
```{r}
selected_genes <- group_by(data_mean, ensembl_id) %>%
  summarize(max = max(mean)) %>%
  filter(max > log(0.1+1))

print(paste0(nrow(selected_genes), " genes with FPKM > 0.01 in at least one age group were selected out of ", length(unique(data_mean$ensembl_id)), " genes in total."))

data_filtered <- filter(data_mean, ensembl_id %in% selected_genes$ensembl_id) 
```


## Select 2000 most variable genes 
```{r}
var <- group_by(data_filtered, ensembl_id) %>%
  # get var per gene over all ages
  summarize(variance = var(mean)) %>%
  arrange(desc(variance)) %>%
  top_n(2000)
head(var)
```

```{r}
filtered_data <- filter(data_filtered, ensembl_id %in% var$ensembl_id) 
```

# MEFISTO
## Create MEFISTO object
```{r}
df_mefisto <- subset(filtered_data, select = c("age", "ensembl_id", "mean")) %>%
  mutate(view = "single_view") 
colnames(df_mefisto) <- c("sample", "feature", "value", "view")
head(df_mefisto)

# df with the corresponding covariate (age) for each sample 
time <- data.frame(sample = age, covariate = "age", value =  age)
```

```{r}
mefisto <- create_mofa(df_mefisto)
mefisto <- set_covariates(mefisto, covariates = time)
print(mefisto)
```

## Train MEFISTO with 1-5 factors
```{r eval=FALSE}
for (n_factors in seq(1,5,1)) {
  data_opts <- get_default_data_options(mefisto)
  
  model_opts <- get_default_model_options(mefisto)
  model_opts$num_factors <- n_factors
  
  train_opts <- get_default_training_options(mefisto)
  
  mefisto_opts <- get_default_mefisto_options(mefisto)
  
  mefisto <- prepare_mofa(mefisto, model_options = model_opts,
                     mefisto_options = mefisto_opts,
                     training_options = train_opts,
                     data_options = data_opts)
  
  outfile = (paste0(file_path, "models/MEFISTO_", n_factors, "f.hdf5"))
  trained_model <- run_mofa(mefisto, outfile)
}
```


## Load trained models
```{r warning= FALSE}
model_1 <- load_model(paste0(file_path, "models/MEFISTO_1f.hdf5"))
model_2 <- load_model(paste0(file_path, "models/MEFISTO_2f.hdf5"))
model_3 <- load_model(paste0(file_path, "models/MEFISTO_3f.hdf5"))
model_4 <- load_model(paste0(file_path, "models/MEFISTO_4f.hdf5"))
model_5 <- load_model(paste0(file_path, "models/MEFISTO_5f.hdf5"))
```


# Model comparison

## Get loadings of the first factor
```{r}
# function to get loadings of the first factor
get_loadings_f1 <- function(model, k){
  weights <- get_weights(model, views = "all", factors = "Factor1", scale = F, as.data.frame = TRUE) %>%
    mutate(value = value/max(abs(value)))
  colnames(weights) <- c("ensembl_id", "factor", "value", "view")
  weights <- weights[order(abs(weights$value), decreasing = TRUE), ] %>%
    left_join(subset(gene_annotations, select = c(ensembl_id, symbol, description))) %>%
    subset(select = c(ensembl_id, value)) %>%
    mutate(n_factors = paste0("k=",k)) %>%
    distinct()
}

loadings <- bind_rows(get_loadings_f1(model_1, 1), get_loadings_f1(model_2, 2),
                      get_loadings_f1(model_3, 3), get_loadings_f1(model_4, 4),
                      get_loadings_f1(model_5, 5)) %>%
  spread(key = n_factors, value = value)
head(loadings)
```


## Correlations between the loadings
```{r}
loadings_cor <- loadings %>% subset(select = -ensembl_id) %>%
  correlate(diagonal = 1)

head(loadings_cor)
```





```{r factors_time}
plot_factors_vs_cov(model_5) + 
  geom_line()+
  theme_bw()+ 
  theme(text = element_text(size = 15))
```


