---
title: "MEFISTO on count values"
author: "Jana"
date: '2022-05-26'
output: BiocStyle::html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Set path
file_path <- "/Users/janabraunger/Documents/Studium/Master/Boston/Masterarbeit/MOFA/"

# Set options for the figures
figdir = paste0(file_path, "Figures/")
if(!dir.exists(figdir)) dir.create(figdir)
knitr::opts_chunk$set(dev = c("png", "pdf"), fig.path = figdir)

# Not use exponential representation of numbers
options(scipen = 999) 
```

```{r load_libraries, message = FALSE, warning = FALSE}
library(tidyverse)
library(magrittr )
library(ggplot2)
library(ggfortify)
library(MOFA2)
library(dplyr)
library(MOFAdata)
library(GGally)
library(ggpubr)
library(pheatmap)
library(RColorBrewer)
library(biomaRt)
library(glmnet)
library(ggvenn)
library(VennDiagram)
library(UpSetR)
```


# Data preprocessing
## RNAseq data
```{r load_data}
data <- read.delim(paste0(file_path, "Data/vst_counts.csv"), 
                                     sep = ",", header = TRUE)
head(data)
```


## Visualize number of individuals per age
```{r hist_ages, fig.height=4}
samples <- subset(data, select = c(names, Age, age_group)) %>%
  distinct()
print(paste0('The dataset includes RNAseq samples from ', nrow(samples), ' patients.'))

ggplot(samples, aes(x = age_group)) +
  geom_bar() +
  theme_bw() + 
  ylab('number of individuals') +
  theme(text = element_text(size = 17))
```


## Mean over all samples in the same age group
```{r}
# get mean expression per age
data_mean <- group_by(data, ensembl_id,gene, age_group) %>%
  summarize(mean = mean(counts)) #%>%
  #mutate(age_group = as.integer(age_group)) %>%
  #sort rows by age
  #arrange(age_group) 
head(data_mean)
age <- unique(data_mean$age_group)
```


## Select 2000 most variable genes 
```{r}
var <- group_by(data_mean, ensembl_id) %>%
  # get var per gene over all ages
  summarize(variance = var(mean)) %>%
  arrange(desc(variance)) %>%
  top_n(2000)
head(var)
```

```{r}
filtered_data <- filter(data_mean, ensembl_id %in% var$ensembl_id) 
```


## PCA as a quality control
```{r pca}
data_wide <- spread(filtered_data, key = age_group, value = mean)
head(data_wide)

#groups <- c("1-9", "10-29", "30-59", "60-79", 
 #                "80-89", "90-96")
groups <- c('1-15', '16-26', '27-60', '61-85', '86-96')

df_pca <- ungroup(data_wide)%>%
  subset(select = -c(ensembl_id, gene))%>%
  t() %>%
  data.frame()%>%
  rownames_to_column(var = "age_group") %>%
  mutate(age_group = factor(age_group, levels = groups))

pca <- prcomp(df_pca[, 2:ncol(df_pca)])
autoplot(pca, data = df_pca, colour = "age_group")  + 
  theme_bw() + 
  theme(text = element_text(size = 20))
```

# MEFISTO
## Create MEFISTO object
```{r}
df_mefisto <- subset(filtered_data, select = c("age_group", "ensembl_id", "mean")) %>%
  mutate(view = "single_view") 
colnames(df_mefisto) <- c("sample", "feature", "value", "view")
head(df_mefisto)

# df with the corresponding covariate (age) for each sample 
time <- data.frame(sample = age, covariate = "age", value = seq(1,5,1))
```

```{r}
mefisto <- create_mofa(df_mefisto)
mefisto <- set_covariates(mefisto, covariates = time)
print(mefisto)
```

## Define options
```{r}
data_opts <- get_default_data_options(mefisto)

model_opts <- get_default_model_options(mefisto)
model_opts$num_factors <- 3

train_opts <- get_default_training_options(mefisto)

mefisto_opts <- get_default_mefisto_options(mefisto)

mefisto <- prepare_mofa(mefisto, model_options = model_opts,
                   mefisto_options = mefisto_opts,
                   training_options = train_opts,
                   data_options = data_opts)
```


## Build and train the MOFA object
```{r}
outfile = (paste0(file_path, "models/test_model_3f.hdf5"))
trained_model <- run_mofa(mefisto, outfile)
```

# Downstream Analysis
```{r variance_explained}
plot_variance_explained(trained_model)
```

```{r}
get_scales(trained_model)
```

```{r factors_time}
plot_factors_vs_cov(trained_model) + 
  geom_line()+
  theme_bw()+ 
  theme(text = element_text(size = 25))
```


```{r heatmap_genes_highest_weight, fig.width=10}
plot_data_heatmap(trained_model,
  view = "single_view",         # view of interest
  factor = 1,             # factor of interest
  features = 15,          # number of features to plot (they are selected by weight)
  
  # extra arguments that are passed to the `pheatmap` function
  cluster_rows = TRUE, cluster_cols = TRUE,
  show_rownames = TRUE, show_colnames = TRUE, fontsize = 15) 
```


# GSEA with MSigDB gene sets
## Run GSEA on weights with positive sign
```{r }
#utils::data("MSigDB_v6.0_C5_human") 
utils::data('reactomeGS')
enrichment.parametric <- run_enrichment(trained_model,
  view = "all", factors = 1:2,
  feature.sets = reactomeGS,
  sign = "positive",
  statistical.test = "parametric"
)
```


**Factor 1**
```{r gsea_f1_pos}
plot_enrichment(enrichment.parametric, 
  factor = 1, 
  max.pathways = 5) +
  theme_classic(base_size = 15)
```

## Run GSEA on weights with negative sign
```{r }
enrichment.parametric <- run_enrichment(trained_model,
  view = "all", factors = 1:2,
  feature.sets = reactomeGS,
  sign = "negative",
  statistical.test = "parametric"
)
```


**Factor 1**
```{r gsea_f1_neg}
plot_enrichment(enrichment.parametric, 
  factor = 1, 
  max.pathways = 5) +
  theme_classic(base_size = 15)
```


# Genes with highest weights
## Get weight matrix
```{r}
weights <- get_weights(trained_model, views = "all", factors = "all", scale = F, as.data.frame = TRUE) %>%
  group_by(factor, view) %>% 
  mutate(value = value/max(abs(value))) %>% 
  ungroup() 
colnames(weights) <- c("ensembl_id", "factor", "value", "view")
weights <- weights[order(abs(weights$value), decreasing = TRUE), ] %>%
  left_join(subset(data, select = c(ensembl_id, gene)) %>% distinct())
head(weights)
```

## Genes with highest weights in factor 1
**Select 10 genes with highest absolute weight in factor 1**
```{r}
selected_genes_pos <- filter(weights, factor == "Factor1") %>%
  top_n(5, value)
selected_genes_neg <- filter(weights, factor == "Factor1") %>%
  top_n(5, -value)
selected_genes <- bind_rows(selected_genes_pos, selected_genes_neg)

selected_genes
```

**Plot expression of selected genes during aging**
```{r f1_expression_over_time, fig.width=10, fig.height = 5}
df_plot = data %>%
  filter(ensembl_id %in% selected_genes$ensembl_id) %>%
  mutate(gene = factor(gene, levels = selected_genes$gene)) 

ggplot(df_plot, aes(x = Age, y = counts)) + 
    geom_point() + 
    #geom_line() +
    facet_wrap(~ gene, scales = "free", ncol = 5) +
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
    labs(y = "vst gene expression counts")  +
    theme_bw() + 
    theme(text = element_text(size = 15)) +
    geom_smooth(color = "red")
```

MT2A: "MT2A is crucial for the immune efficiency during aging and age-related diseases" https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5037761/ 

SVEP1: "SVEP1 has emerged as a novel circulating biomarker of age and longevity in humans" https://longerlife.org/research/stitziel-targeting-receptor-interactions-with-svep1-261 


**Get chromosome positions**
```{r}
mart <- useEnsembl(biomart = "ensembl", 
                   dataset = "hsapiens_gene_ensembl", 
                   mirror = "uswest")
genes <- data$Transcript.ID
gene_annotations <- getBM(filters="external_gene_name", 
                          attributes=c("external_gene_name", 
                                       "chromosome_name", "start_position", 
                                       "end_position", "description"), 
                          values=unique(weights$gene), mart=mart) %>%
  data.frame() %>%
  arrange(chromosome_name) %>%
  group_by(external_gene_name) %>%
  summarize(chr = dplyr::first(chromosome_name), start = dplyr::first(start_position), 
            end = dplyr::first(end_position), description = dplyr::first(description)) %>%
  dplyr::rename(gene = external_gene_name)

head(gene_annotations)
```


**Save 10% of genes with highest positive and negative weight in factor 1**
```{r}
weights_f1 <- filter(weights, factor == "Factor1") %>%
  group_by(gene) %>%
  # in case multiple ensembl ids map to the same gene symbol the max loading is kept
  summarize(value = max(value))

# Select genes with highest positive weight
top_genes_pos <- weights_f1 %>%
  top_frac(0.1, value) %>%
  left_join(gene_annotations, by = "gene") %>%
  mutate(locus = paste0("chr_", chr, "_loc_", 
                        plyr::round_any(start, 250000, floor))) %>%
  subset(select = c(gene, value, locus))

# Select genes with highest negative weight
top_genes_neg <- weights_f1 %>%
  top_frac(0.1, -value) %>%
  left_join(gene_annotations, by = "gene") %>%
  mutate(locus = paste0("chr_", chr, "_loc_", 
                        plyr::round_any(start, 250000, floor))) %>%
  subset(select = c(gene, value, locus))

top_genes <- bind_rows(top_genes_pos, top_genes_neg)
head(top_genes)

# Save as csv
write.csv(top_genes_pos, paste0(file_path, "Data/top_genes_pos_0.1.csv"), row.names = FALSE)
write.csv(top_genes_neg, paste0(file_path, "Data/top_genes_neg_0.1.csv"), row.names = FALSE)
```


**Distribution of weights in factor 1**

```{r hist_weights}
weights_f1$color <- factor(ifelse(weights_f1$gene %in% top_genes$gene, 
                                  'selected_genes', 'others'), 
                           levels = c('selected_genes', 'others'))
ggplot(weights_f1, aes(x = value, fill = color)) + 
  geom_histogram(bins = 100) +
  theme_bw() + 
  theme(text = element_text(size = 15))
```

```{r}
# Number of genes with positive weight
length(unique(filter(weights_f1, value > 0)$gene))
# Number of genes with negative weight
length(unique(filter(weights_f1, value < 0)$gene))
```



## Comparison with active loci in IMR90 and GM23248
```{r}
IMR90 <- read.csv(paste0(file_path, "Data/IMR90_cosine_wo_PC.txt")) %>%
  dplyr::rename(locus = X0) %>%
  mutate(young_activity = "young-active")
old_fibroblasts <- read.csv(paste0(file_path, "Data/old_fibroblasts_cosine_wo_PC.txt")) %>%
  dplyr::rename(locus = X0) %>%
  mutate(old_activity = "old-active")

# add for each gene whether corresponding loci is active or inactive based on the HiC clustering in IMR90 and GM23248
activity <- top_genes_pos %>%
  mutate(sign = 'positive') %>%
  bind_rows(top_genes_neg %>% mutate(sign = 'negative')) %>%
  mutate(sign = factor(sign, levels = c('positive', 'negative'))) %>%
  left_join(IMR90) %>% 
  replace_na(list(young_activity = "young-inactive")) %>%
  left_join(old_fibroblasts) %>% 
  replace_na(list(old_activity = "old-inactive")) %>%
  mutate(development = factor(paste0(young_activity, "_", old_activity), 
        levels = c("young-active_old-active", "young-active_old-inactive", "young-inactive_old-active", "young-inactive_old-inactive")))
head(activity)
```


```{r development_activity, fig.width = 8}
ggplot(activity, aes(x = development, fill = development)) +
  geom_bar() +
  facet_wrap(~sign) +
  theme_bw() +
  theme(axis.text.x=element_blank(), 
        axis.ticks.x=element_blank()) + 
  theme(text = element_text(size = 17))
```


## Linear regression
**Linear regression of top 10 genes**
```{r}
selected_genes <- filter(weights, factor == "Factor1") %>%
  top_n(10, abs(value))

df = data %>%
  filter(ensembl_id %in% selected_genes$ensembl_id) %>%
  subset(select = c(names, gene, Age, counts)) %>%
  spread(key = gene, value = counts)

model <- lm(Age ~ ., subset(df, select = -names))
summary(model)
```


**LASSO**
```{r}
df = data %>%
  subset(select = c(names, Age, ensembl_id, counts)) %>%
  filter(ensembl_id %in% filtered_data$ensembl_id) %>%
  spread(key = ensembl_id, value = counts)

model <- glmnet(subset(df, select = -c(names, Age)), df$Age, alpha = 1, lambda = 1)
lasso <- data.frame(ensembl_id = dimnames(coef(model))[[1]], coefficient = matrix(coef(model))) %>%
  filter(coefficient != 0) %>%
  left_join(subset(data, select = c(ensembl_id, gene)) %>% distinct()) %>%
  filter(ensembl_id != "(Intercept)")
#write.csv(lasso, paste0(file_path, "Data/lasso_coef.csv"), row.names = FALSE)
head(lasso)
```


```{r}
df %<>% mutate(predicted_age = lapply(1:nrow(df), function(row){
  stats::predict(model, s = 0.1, newx = as.matrix(df[row, 3:ncol(df)]))}) %>% 
    unlist())

ggplot(df, aes(x = Age, y = predicted_age)) +
  geom_point() + theme_bw()

pred_age = subset(df, select = c('names', 'predicted_age')) %>%
  mutate(predicted_age = as.character(as.integer(predicted_age)))
```



```{r venn_pos}
ggvenn(list(pos_loadings = top_genes_pos$gene, pos_lasso = filter(lasso, coefficient > 0)$gene)) + 
    theme(text = element_text(size = 15))
```


```{r venn_neg}
ggvenn(list(neg_loadings = top_genes_neg$gene, neg_lasso = filter(lasso, coefficient < 0)$gene)) + 
    theme(text = element_text(size = 15))
```

```{r upset_lasso}
list = list(MEFISTO_pos = top_genes_pos$gene, lasso_pos = filter(lasso, coefficient > 0)$gene,
            MEFISTO_neg = top_genes_neg$gene, lasso_neg = filter(lasso, coefficient < 0)$gene)

upset(fromList(list), order.by = "freq", nsets = 4,
       mainbar.y.label = "Intersections", sets.x.label = "DE Genes",
       text.scale = c(2.5, 2.5, 2, 2, 2.5, 2.5), nintersects = 10)
```


## MEFISTO factors instead of PCA for sorting individuals
```{r}
weights_f1 <- filter(weights, factor == "Factor1") %>%
  subset(select = c(ensembl_id, gene, value)) %>%
  dplyr::rename("weight_f1" = "value")

weights_f2 <- filter(weights, factor == "Factor2") %>%
  subset(select = c(ensembl_id, gene, value)) %>%
  dplyr::rename("weight_f2" = "value")

MEFISTO_scores <- left_join(data, weights_f1, 
                            by = c("ensembl_id", "gene")) %>%
  left_join(weights_f2, by = c("ensembl_id", "gene")) %>%
  drop_na() %>%
  mutate(score1 = counts * weight_f1) %>%
  mutate(score2 = counts * weight_f2) %>%
  group_by(names, Age, age_group) %>%
  summarize(score1 = sum(score1), score2 = sum(score2))

head(MEFISTO_scores)
```


```{r}
ggplot(MEFISTO_scores, aes(x = score1, y = score2, color = age_group)) + geom_point()
ggplot(MEFISTO_scores, aes(x = Age, y = score1, color = age_group)) + geom_point()
ggplot(MEFISTO_scores, aes(x = age_group, y = score1, fill = age_group)) + geom_boxplot()
```


# Rank TFs based on MEFISTO loadings
```{r}
# Load TF-target interactions
TF_targets <- read.delim(paste0(dir, "Data/tf-target-information.txt")) %>%
  subset(select = c("TF", "target")) %>%
  distinct() %>%
  filter(target %in% filtered_data$gene)

# Filter for TFs that have more than 5 targets (otherwise some with one target dominate the whole plot)
n_targets = group_by(TF_targets, TF) %>%
  summarize(n = n()) %>% 
  filter(n > 5)

TF_targets %<>% filter(TF %in% n_targets$TF) %>%
  dplyr::rename("gene" = "target")

head(TF_targets)
```

```{r}
# Calculate mean absolute loading of the targets per TF
mean_loadings <- TF_targets %>% 
  # add loading in factor 1 for each target gene
  inner_join(filter(weights, factor == "Factor1") %>% 
               subset(select = c(ensembl_id, gene, value))) %>% 
  # mean over absolute loadings per TF
  group_by(TF) %>%
  summarize(value = median(abs(value))) 
head(mean_loadings)
```


```{r}
# add info about which TFs occurred in which of the three Steiner trees
pcst_TFs = read.delim(paste0(dir, 'Data/pcst/incl_TFs_design2.csv'), sep = ",") %>%
  mutate(included = "True") %>%
  mutate(net = factor(net, levels = c('young_net', 'middle_net', 'old_net'))) %>% #keeps ordering
  spread(key = net, value = included, fill = "False") %>%
  right_join(mean_loadings) %>%
  replace(is.na(.), 'False') %>%
  arrange(desc(value)) 
pcst_TFs %<>% mutate(TF = factor(TF, levels = unique(pcst_TFs$TF)))
head(pcst_TFs)
```

```{r}
ggplot(pcst_TFs, aes(x = TF)) +
  geom_point(aes(x = TF, y = value)) +
  geom_vline(xintercept = pcst_TFs[pcst_TFs$old_net == 'True', 'TF'], color = "red")

ggplot(pcst_TFs, aes(x = TF)) +
  geom_point(aes(x = TF, y = value)) +
  geom_vline(xintercept = TFs, color = "red")
```

```{r}
# Select TFs that occur in all 3 Steiner trees in design 2
TFs = c('AHR','ATF1','BACH2','FLI1','FOXO3','GATA4','GTF2B','HIF1A','KAT5',
        'LYL1','MAFF','MAZ','NOTCH1','NR2C2','SRC','STAT1','STAT3','TCF7L2','TEAD4')

filter(pcst_TFs, TF %in% TFs) %>%
  slice_max(n = 10, order_by = value) %>%
  subset(select = c(TF, value))
```


STAT3: https://genomics.senescence.info/genes/entry.php?hgnc=STAT3
BACH2: marker of DNA damage and ageing (https://pubmed.ncbi.nlm.nih.gov/24075570/)
NOTCH: https://pubmed.ncbi.nlm.nih.gov/27328278/

