{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cc3be39d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import numpy as np \n",
    "import pandas as pd\n",
    "import pybedtools"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77f96a62",
   "metadata": {},
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c7beca1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_dir = '/home/braunger/masterthesis/data/genome_data/'\n",
    "epigenome_dir = '/home/braunger/masterthesis/data/regulatory_data/regulatory_data_old_fibroblasts/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "611bbd75",
   "metadata": {},
   "source": [
    "Data about the positions of the transcripts on the chromosomes were downloaded using the UCSC Table Browser (assembly: hg19, group: Genes and Gene Predictions, track: Ensembl Genes, table: ensGene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9c30fa54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000237247</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>66999065</td>\n",
       "      <td>67210057</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000371039</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>66999274</td>\n",
       "      <td>67210768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000424320</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>66999297</td>\n",
       "      <td>67145425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000371035</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>66999822</td>\n",
       "      <td>67208882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000468286</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>66999838</td>\n",
       "      <td>67142779</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              name chrom strand   txStart     txEnd\n",
       "0  ENST00000237247  chr1      +  66999065  67210057\n",
       "1  ENST00000371039  chr1      +  66999274  67210768\n",
       "2  ENST00000424320  chr1      +  66999297  67145425\n",
       "3  ENST00000371035  chr1      +  66999822  67208882\n",
       "4  ENST00000468286  chr1      +  66999838  67142779"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_annotations = pd.read_csv(genome_dir+'gene_annotation.tsv', sep = '\\t', header=0)\n",
    "selected_cols = [\"name\", \"chrom\", \"strand\", \"txStart\", \"txEnd\"]\n",
    "transcript_annotations = transcript_annotations.loc[:, selected_cols]\n",
    "\n",
    "transcript_annotations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5edb83e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>624</th>\n",
       "      <td>ENST00000373020</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>631</th>\n",
       "      <td>ENST00000371584</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>632</th>\n",
       "      <td>ENST00000371588</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>639</th>\n",
       "      <td>ENST00000423670</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>642</th>\n",
       "      <td>ENST00000359326</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                name  count\n",
       "624  ENST00000373020     12\n",
       "631  ENST00000371584      1\n",
       "632  ENST00000371588     22\n",
       "639  ENST00000423670      3\n",
       "642  ENST00000359326      1"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_seq_counts = pd.read_csv(epigenome_dir + 'GSM2072585_ENCFF913ZKI_transcript_quantifications_hg19.tsv', sep = '\\t', header = 0)\n",
    "#select columns\n",
    "selected_cols = [\"transcript_id\", \"FPKM\"]\n",
    "rna_seq_counts = rna_seq_counts.loc[:, selected_cols]\n",
    "#convert counts to ints\n",
    "rna_seq_counts[\"FPKM\"] = rna_seq_counts[\"FPKM\"].astype(int)\n",
    "#remove zeroes\n",
    "rna_seq_counts = rna_seq_counts[rna_seq_counts.FPKM > 0]\n",
    "#remove version number of transcript_id\n",
    "rna_seq_counts['transcript_id'] = rna_seq_counts['transcript_id'].str.split(r'.').str.get(0)\n",
    "#rename columns\n",
    "rna_seq_counts.columns = [\"name\", \"count\"]\n",
    "\n",
    "rna_seq_counts.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "749133bc",
   "metadata": {},
   "source": [
    "## Merge into one data frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7dc981de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>txStart</th>\n",
       "      <th>txEnd</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000373020_1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>99883666</td>\n",
       "      <td>99891803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000373020_2</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>99883666</td>\n",
       "      <td>99891803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000373020_3</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>99883666</td>\n",
       "      <td>99891803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000373020_4</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>99883666</td>\n",
       "      <td>99891803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000373020_5</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>99883666</td>\n",
       "      <td>99891803</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                name chrom strand   txStart     txEnd\n",
       "0  ENST00000373020_1  chrX      -  99883666  99891803\n",
       "1  ENST00000373020_2  chrX      -  99883666  99891803\n",
       "2  ENST00000373020_3  chrX      -  99883666  99891803\n",
       "3  ENST00000373020_4  chrX      -  99883666  99891803\n",
       "4  ENST00000373020_5  chrX      -  99883666  99891803"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Merge the two data frames\n",
    "rna = pd.merge(rna_seq_counts, transcript_annotations, on = \"name\")\n",
    "\n",
    "#Duplicate rows according to the count number\n",
    "rna_long = rna.loc[rna.index.repeat(rna.pop('count'))]\n",
    "rna_long['name'] = rna_long['name'].astype(str) + '_' + rna_long.groupby(level=0).cumcount().add(1).astype(str)\n",
    "rna_long = rna_long.reset_index(drop=True)\n",
    "\n",
    "rna_long.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d350dc41",
   "metadata": {},
   "source": [
    "## Convert to bed format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "78ac2ad9",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_long = rna_long[[\"chrom\", \"txStart\", \"txEnd\", \"name\", \"strand\"]]\n",
    "rna_long.columns = [\"chrom\", \"start\", \"end\", \"name\", \"score\"]\n",
    "rna_bed = pybedtools.BedTool.from_dataframe(rna_long)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ee07f2cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/home/braunger/masterthesis/data/regulatory_data/regulatory_data_old_fibroblasts/rna-seq.bed)>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_bed.saveas(epigenome_dir + \"rna-seq.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ea60e9fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>99883666</td>\n",
       "      <td>99891803</td>\n",
       "      <td>ENST00000373020_1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>99883666</td>\n",
       "      <td>99891803</td>\n",
       "      <td>ENST00000373020_2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>99883666</td>\n",
       "      <td>99891803</td>\n",
       "      <td>ENST00000373020_3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>99883666</td>\n",
       "      <td>99891803</td>\n",
       "      <td>ENST00000373020_4</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>99883666</td>\n",
       "      <td>99891803</td>\n",
       "      <td>ENST00000373020_5</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5911399</th>\n",
       "      <td>chr19</td>\n",
       "      <td>58790317</td>\n",
       "      <td>58807254</td>\n",
       "      <td>ENST00000608843_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5911400</th>\n",
       "      <td>chr19</td>\n",
       "      <td>58790317</td>\n",
       "      <td>58807254</td>\n",
       "      <td>ENST00000608843_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5911401</th>\n",
       "      <td>chr19</td>\n",
       "      <td>58790317</td>\n",
       "      <td>58807254</td>\n",
       "      <td>ENST00000608843_3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5911402</th>\n",
       "      <td>chr4</td>\n",
       "      <td>164450922</td>\n",
       "      <td>164451849</td>\n",
       "      <td>ENST00000609356_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5911403</th>\n",
       "      <td>chr4</td>\n",
       "      <td>164450922</td>\n",
       "      <td>164451849</td>\n",
       "      <td>ENST00000609356_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5911404 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chrom      start        end               name score\n",
       "0         chrX   99883666   99891803  ENST00000373020_1     -\n",
       "1         chrX   99883666   99891803  ENST00000373020_2     -\n",
       "2         chrX   99883666   99891803  ENST00000373020_3     -\n",
       "3         chrX   99883666   99891803  ENST00000373020_4     -\n",
       "4         chrX   99883666   99891803  ENST00000373020_5     -\n",
       "...        ...        ...        ...                ...   ...\n",
       "5911399  chr19   58790317   58807254  ENST00000608843_1     +\n",
       "5911400  chr19   58790317   58807254  ENST00000608843_2     +\n",
       "5911401  chr19   58790317   58807254  ENST00000608843_3     +\n",
       "5911402   chr4  164450922  164451849  ENST00000609356_1     +\n",
       "5911403   chr4  164450922  164451849  ENST00000609356_2     +\n",
       "\n",
       "[5911404 rows x 5 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "280ea207",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
